pkg/checksum/checker.go

```diff
@@ -229,8 +230,75 @@ func (c *Checker) replaceChunk(chunk *table.Chunk) error {
+	//
+	// ------------------------
+	// LATEST DETECTED DEADLOCK
+	// ------------------------
+	// 2024-06-11 18:34:21 70676106989440
+	// *** (1) TRANSACTION:
+	// TRANSACTION 15106308424, ACTIVE 4 sec updating or deleting
+	// mysql tables in use 1, locked 1
+	// LOCK WAIT 620 lock struct(s), heap size 73848, 49663 row lock(s), undo log entries 49661
+	// MySQL thread id 540806, OS thread handle 70369444421504, query id 409280999 10.137.84.232 <snip> updating
+	// DELETE FROM `<snip>`.`_<snip>_new` WHERE `id` >= 1108588365 AND `id` < 1108688365
+	//
+	// *** (1) HOLDS THE LOCK(S):
+	// RECORD LOCKS space id 1802 page no 26277057 n bits 232 index idempotence_key_idx of table `<snip>`.`_<snip>_new` trx id 15106308424 lock_mode X locks rec but not gap
+	// Record lock, heap no 163 PHYSICAL RECORD: n_fields 2; compact format; info bits 32
+	// 0: len 30; hex <snip>; asc <snip>; (total 62 bytes);
+	// 1: len 8; hex <snip>; asc     <snip>;;
+	//
+	//
+	// *** (1) WAITING FOR THIS LOCK TO BE GRANTED:
+	// RECORD LOCKS space id 1802 page no 1945840 n bits 280 index idempotence_key_idx of table `<snip>`.`_<snip>_new` trx id 15106308424 lock_mode X locks rec but not gap waiting
+	// Record lock, heap no 75 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
+	// 0: len 30; hex <snip>; asc <snip>; (total 62 bytes);
+	// 1: len 8; hex <snip>; asc     <snip>;;
+	//
+	//
+	// *** (2) TRANSACTION:
+	// TRANSACTION 15106301192, ACTIVE 58 sec inserting
+	// mysql tables in use 2, locked 1
+	// LOCK WAIT 220020 lock struct(s), heap size 27680888, 409429 row lock(s), undo log entries 162834
+	// MySQL thread id 540264, OS thread handle 70369485823872, query id 409127061 10.137.84.232 <snip> executing
+	// REPLACE INTO `<snip>`.`_<snip>_new` (`id`, <snip> FROM `<snip>`.`<snip>` WHERE `id` >= 1106488365 AND `id` < 1106588365
+	//
+	// *** (2) HOLDS THE LOCK(S):
+	// RECORD LOCKS space id 1802 page no 1945840 n bits 280 index idempotence_key_idx of table `<snip>`.`_<snip>_new` trx id 15106301192 lock_mode X
+	// Record lock, heap no 75 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
+	// 0: len 30; hex <snip>; asc <snip>; (total 62 bytes);
+	// 1: len 8; hex <snip>; asc     B yI;;
+	//
+	//
+	// *** (2) WAITING FOR THIS LOCK TO BE GRANTED:
+	// RECORD LOCKS space id 1802 page no 26277057 n bits 232 index idempotence_key_idx of table `<snip>`.`_<snip>_new` trx id 15106301192 lock_mode X waiting
+	// Record lock, heap no 163 PHYSICAL RECORD: n_fields 2; compact format; info bits 32
+	// 0: len 30; hex <snip>; asc <snip>; (total 62 bytes);
+	// 1: len 8; hex <snip>; asc     <snip>;;
+	//
+	// *** WE ROLL BACK TRANSACTION (1)
+	//
+	// We don't need this to be an atomic transaction. We just need to delete from the _new table
+	// first so that any since-deleted rows (which wouldn't get removed by replace) are removed first.
+	// By doing this as two transactions we should be able to remove
+	// the opportunity for deadlocks.
```

Comment:
Can deadlock happen with multiple sequential stmts in same txn?

 Is deadlock mainly because of parallel chunk recopying(across multiple txns) instead of single chunk delete and replace in a single txn?

If we have a mutex to avoid parallel copy of chunks, do we still need to split into multiple txns for a single chunk?

Reply:
> Can deadlock happen with multiple sequential stmts in same txn?
No, it shouldn't be possible to deadlock with itself. In the case identified it was two independent transactions from two different re-copy tasks.

> Is deadlock mainly because of parallel chunk recopying(across multiple txns) instead of single chunk delete and replace in a single txn?

I believe it may be a good enough fix to split delete and replace. But because we have no requirement to be same transaction I implemented both fixes.

> If we have a mutex to avoid parallel copy of chunks, do we still need to split into multiple txns for a single chunk?

It's more so that there is no requirement to be one transaction, and in future we may use parallelism again, but I don't think it's likely.